toolkit/tools/imagegen/configuration/packagerepo.go

```diff
@@ -0,0 +1,217 @@
+	}
+	return
+}
+
+// IsValid returns an error if the PackageRepo struct is not valid
+func (p *PackageRepo) IsValid() (err error) {
+	err = p.nameIsValid()
+	if err != nil {
+		return
+	}
+
+	err = p.repoUrlIsValid()
+	if err != nil {
+		return
+	}
+
+	return
+}
+
+// UpdatePackageRepo creates additional repo files specified by image configuration
+// and returns error if the operation fails
+func UpdatePackageRepo(installChroot *safechroot.Chroot, config SystemConfig) (err error) {
+	const (
+		repoFileDir   = "/etc/yum.repos.d/"
+		localRepoFile = "/etc/yum.repos.d/mariner-iso.repo"
+		squashErrors  = false
+	)
+
+	if exists, ferr := file.DirExists(repoFileDir); ferr != nil {
+		logger.Log.Errorf("Error accessing repo file directory %s.", repoFileDir)
+		err = ferr
+		return
+	} else if !exists {
+		err = fmt.Errorf("Could not find the repo file directory %s to update package repo", repoFileDir)
+		return
+	}
+
+	// Remove mariner-iso.repo
+	os.Remove(localRepoFile)
+
+	// Loop through the PackageRepos field to determine if any customized package repos are specified.
+	// If specified, create new repo files for them
+	for _, packageRepo := range config.PackageRepos {
+		err = createCustomPackageRepo(installChroot, packageRepo, repoFileDir)
+		if err != nil {
+			return
+		}
+	}
+
+	// It is possible that network access may not be up at this point,
+	// so check network access
+	err, hasNetworkAccess := network.CheckNetworkAccess()
+	if !hasNetworkAccess {
+		err = fmt.Errorf("no network access in the system")
+	}
+	return
+}
+
+// nameIsValid returns an error if the package repo name is empty
+func (p *PackageRepo) nameIsValid() (err error) {
+	if strings.TrimSpace(p.Name) == "" {
+		return fmt.Errorf("invalid value for package repo name (%s), name cannot be empty", p.Name)
+	}
+	return
+}
+
+// repoUrlIsValid returns an error if the package url is invalid
+func (p *PackageRepo) repoUrlIsValid() (err error) {
+	if strings.TrimSpace(p.BaseUrl) == "" {
+		return fmt.Errorf("invalid value for package repo URL (%s), URL cannot be empty", p.BaseUrl)
+	}
+
+	_, err = url.ParseRequestURI(p.BaseUrl)
+	if err != nil {
+		logger.Log.Errorf("Failed to parse input URL %s, Error: %s", p.BaseUrl, err)
+	}
+	return
+}
+
+func writeAdditionalFields(stringBuilder *strings.Builder) (err error) {
+	const (
+		gpgKey       = "gpgkey=file:///etc/pki/rpm-gpg/MICROSOFT-RPM-GPG-KEY file:///etc/pki/rpm-gpg/MICROSOFT-METADATA-GPG-KEY\n"
+		enable       = "enabled=1\n"
+		gpgCheck     = "gpgcheck=1\n"
+		repogpgCheck = "repo_gpgcheck=1\n"
+		skip         = "skip_if_unavailable=True\n"
+		sslVerify    = "sslverify=1\n"
+	)
+
+	additionalFields := []string{gpgKey, enable, gpgCheck, repogpgCheck, skip, sslVerify}
+
+	for idx := range additionalFields {
+		_, err = stringBuilder.WriteString(additionalFields[idx])
+		if err != nil {
+			logger.Log.Errorf("Error writing additional fields: %s. Error: %s", additionalFields, err)
+			return
+		}
+	}
```

Comment:
Nit: no need to use the index, Go already retrieves the value for you:
```suggestion
	for _, additionalField := range additionalFields {
		_, err = stringBuilder.WriteString(additionalField)
		if err != nil {
			logger.Log.Errorf("Error writing additional field '%s' out of all fields: %s. Error: %s", additionalField, additionalFields, err)
			return
		}
	}
```